src/main/java/com/hubspot/jinjava/lib/tag/eager/EagerIfTag.java

```diff
@@ -0,0 +1,131 @@
+package com.hubspot.jinjava.lib.tag.eager;
+
+import com.hubspot.jinjava.interpret.DeferredValueException;
+import com.hubspot.jinjava.interpret.JinjavaInterpreter;
+import com.hubspot.jinjava.interpret.TemplateSyntaxException;
+import com.hubspot.jinjava.lib.tag.ElseIfTag;
+import com.hubspot.jinjava.lib.tag.ElseTag;
+import com.hubspot.jinjava.lib.tag.IfTag;
+import com.hubspot.jinjava.tree.Node;
+import com.hubspot.jinjava.tree.TagNode;
+import com.hubspot.jinjava.util.LengthLimitingStringBuilder;
+import com.hubspot.jinjava.util.ObjectTruthValue;
+import org.apache.commons.lang3.StringUtils;
+
+public class EagerIfTag extends EagerTagDecorator<IfTag> {
+
+  public EagerIfTag() {
+    super(new IfTag());
+  }
+
+  public EagerIfTag(IfTag ifTag) {
+    super(ifTag);
+  }
+
+  @Override
+  public String eagerInterpret(TagNode tagNode, JinjavaInterpreter interpreter) {
+    if (StringUtils.isBlank(tagNode.getHelpers())) {
+      throw new TemplateSyntaxException(
+        interpreter,
+        tagNode.getMaster().getImage(),
+        "Tag 'if' expects expression"
+      );
+    }
+
+    LengthLimitingStringBuilder result = new LengthLimitingStringBuilder(
+      interpreter.getConfig().getMaxOutputSize()
+    );
+
+    result.append(
+      executeInChildContext(
+        eagerInterpreter ->
+          getEagerImage(tagNode.getMaster(), eagerInterpreter) +
+          renderChildren(tagNode, eagerInterpreter),
+        interpreter,
+        false
+      )
+    );
+    tagNode.getMaster().setRightTrimAfterEnd(false);
+    result.append(reconstructEnd(tagNode));
+
+    return result.toString();
+  }
+
+  @Override
+  public String renderChildren(TagNode tagNode, JinjavaInterpreter interpreter) {
+    // If the branch is impossible, it should be removed.
+    boolean definitelyDrop = shouldDropBranch(tagNode, interpreter);
+    // If an ("elseif") branch would definitely get executed,
+    // change it to an "else" tag and drop all the subsequent branches.
+    // We know this has to start as false otherwise IfTag would have chosen
+    // the first branch.
+    boolean definitelyExecuted = false;
+    StringBuilder sb = new StringBuilder();
+    for (Node child : tagNode.getChildren()) {
+      if (TagNode.class.isAssignableFrom(child.getClass())) {
+        TagNode tag = (TagNode) child;
+        if (
+          tag.getName().equals(ElseIfTag.TAG_NAME) ||
+          tag.getName().equals(ElseTag.TAG_NAME)
+        ) {
+          if (definitelyExecuted) {
+            break;
+          }
+          definitelyDrop =
```

Comment:
if we drop a branch, what happens to the whitespace around it? I could see this changing the output of templates because of dropped statements and whitespace (not that it's generally useful).

Reply:
This is a good question. The trimming gets handled when the tree gets parsed. Dropping a statement here is done similarly to picking the correct branch in the normal IfTag (and then not executing any of the others). Since no newlines are being added from the EagerIfTag's logic, the resulting whitespace is accurate. 

The TreeParser takes the trim characters from a TagToken and moves them over to the TextToken children so we the whitespace trimming doesn't get handled in the tag node logic.